### PR TITLE
Prefer const lint rule & fixes

### PR DESCRIPTION
**Pull request by weswigham**
Prefers `const` over `let` when the declared variable never appears on the left hand side of an assignment.

### CODE REVIEW

File: scripts/tslint/preferConstRule.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    declaration: ts.VariableDeclaration;
+    usages: number;
+}
+
+class PreferConstWalker extends Lint.RuleWalker {
+    private inScopeLetDeclarations: ts.Map<DeclarationUsages>[] = [];
+    private errors: Lint.RuleFailure[] = [];
+    private markAssignment(identifier: ts.Identifier) {
+        const name = identifier.text;
+        for (var i = this.inScopeLetDeclarations.length - 1; i >= 0; i--) {
```
Comment (DanielRosenwasser) [inline scripts/tslint/preferConstRule.ts]: `let`

File: scripts/tslint/preferConstRule.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            var declarations = this.inScopeLetDeclarations[i];
```
Comment (DanielRosenwasser) [inline scripts/tslint/preferConstRule.ts]: `const`

File: scripts/tslint/preferConstRule.ts (hunk 1 above)
Comment (weswigham) [inline scripts/tslint/preferConstRule.ts]: :sparkles:

File: scripts/tslint/preferConstRule.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        node = node.parent;
+    }
+
+    return node;
+}
+
+function getCombinedNodeFlags(node: ts.Node): ts.NodeFlags {
+    node = walkUpBindingElementsAndPatterns(node);
+
+    let flags = node.flags;
```
Comment (DanielRosenwasser) [inline scripts/tslint/preferConstRule.ts]: `const`

Comment (weswigham) [inline scripts/tslint/preferConstRule.ts]: That would be an error, good sir. `flags` is mutable and appears on the left hand side of the `|=` operator below.

File: scripts/tslint/preferConstRule.ts (hunk 2 above)
Comment (weswigham) [inline scripts/tslint/preferConstRule.ts]: :+1:

File: scripts/tslint/preferConstRule.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline scripts/tslint/preferConstRule.ts]: My bad :smile:

File: scripts/tslint/preferConstRule.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        super.visitSourceFile(node);
+        // Sort errors by position because tslint doesn't
+        this.errors.sort((a, b) => a.getStartPosition().getPosition() - b.getStartPosition().getPosition()).forEach(e => this.addFailure(e));
+    }
+
+    visitBinaryExpression(node: ts.BinaryExpression) {
+        if (isAssignmentOperator(node.operatorToken.kind)) {
+            this.visitLHSExpressions(node.left);
+        }
+        super.visitBinaryExpression(node);
```
Comment (yuit) [inline scripts/tslint/preferConstRule.ts]: curious why you call the `super.visitBinaryExpression` after ? or it doesn't matter?

Comment (weswigham) [inline scripts/tslint/preferConstRule.ts]: I want my logic to occur before we decend into the expression.

File: scripts/tslint/preferConstRule.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        }
+        else if (isBindingLiteralExpression(node)) {
+            this.visitBindingLiteralExpression(node as (ts.ArrayLiteralExpression | ts.ObjectLiteralExpression));
+        }
+    }
+    
+    private visitBindingLiteralExpression(node: ts.ArrayLiteralExpression | ts.ObjectLiteralExpression) {
+        if (node.kind === ts.SyntaxKind.ObjectLiteralExpression) {
+            const pattern = node as ts.ObjectLiteralExpression;
+            for (let i = 0; i < pattern.properties.length; i++) {
```
Comment (yuit) [inline scripts/tslint/preferConstRule.ts]: nit: use `for...of`

**Comment by DanielRosenwasser** [comment]
Other than @yuit's comment about the `for..of`, :+1:

**Pull request merged**